code/modules/mob/living/basic/space_fauna/revenant/_revenant.dm

```diff
@@ -0,0 +1,513 @@
+	return //most humans will now be either bones or harvesters, but we're still un-alive.
+
+/mob/living/basic/revenant/bullet_act()
+	if(!revealed || dormant)
+		return BULLET_ACT_FORCE_PIERCE
+	return ..()
+
+/mob/living/basic/revenant/proc/on_move(datum/source, atom/entering_loc)
+	SIGNAL_HANDLER
+	if(isnull(orbiting) || incorporeal_move_check(entering_loc))
+		return
+
+	// we're about to go somewhere we aren't meant to, end the orbit and block the move
+	orbiting.end_orbit(src)
+	return COMPONENT_MOVABLE_BLOCK_PRE_MOVE
+
+/// Generates the information the player needs to know how to play their role, and returns it as a list.
+/mob/living/basic/revenant/proc/create_login_string()
+	RETURN_TYPE(/list)
+	var/list/returnable_list = list()
+	returnable_list += span_deadsay(span_boldbig("You are a revenant."))
+	returnable_list += span_bold("Your formerly mundane spirit has been infused with alien energies and empowered into a revenant.")
+	returnable_list += span_bold("You are not dead, not alive, but somewhere in between. You are capable of limited interaction with both worlds.")
+	returnable_list += span_bold("You are invincible and invisible to everyone but other ghosts. Most abilities will reveal you, rendering you vulnerable.")
+	returnable_list += span_bold("To function, you are to drain the life essence from humans. This essence is a resource, as well as your health, and will power all of your abilities.")
+	returnable_list += span_bold("<i>You do not remember anything of your past lives, nor will you remember anything about this one after your death.</i>")
+	returnable_list += span_bold("Be sure to read <a href=\"https://tgstation13.org/wiki/Revenant\">the wiki page</a> to learn more.")
+	return returnable_list
+
+/mob/living/basic/revenant/proc/set_random_revenant_name()
+	var/list/built_name_strings = list()
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "spirit_type"))
+	built_name_strings += " of "
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "adverb"))
+	built_name_strings += pick(strings(REVENANT_NAME_FILE, "theme"))
+	name = built_name_strings.Join("")
+
+/mob/living/basic/revenant/proc/on_baned(obj/item/weapon, mob/living/user)
+	SIGNAL_HANDLER
+	visible_message(
+		span_warning("[src] violently flinches!"),
+		span_revendanger("As [weapon] passes through you, you feel your essence draining away!"),
+	)
+	inhibited = TRUE
+	update_mob_action_buttons()
+	addtimer(CALLBACK(src, PROC_REF(reset_inhibit)), 3 SECONDS)
+
+/mob/living/basic/revenant/proc/reset_inhibit()
+	inhibited = FALSE
+	update_mob_action_buttons()
+
+/// Incorporeal move check: blocked by holy-watered tiles and salt piles.
+/mob/living/basic/revenant/proc/incorporeal_move_check(atom/destination)
+	var/turf/open/floor/step_turf = get_turf(destination)
+	if(isnull(step_turf))
+		return TRUE // what? whatever let it happen
+
+	if(step_turf.turf_flags & NOJAUNT)
+		to_chat(src, span_warning("Some strange aura is blocking the way."))
```

Comment:
Why is salt a balloon alert but the rest of these are `to_chat`?

Reply:
because i couldn't figure out a good way to condense the other two to a `to_chat()` and also salt can be hidden under tables sometimes while holy auras have an overlay iirc